src/map/entities/mobentity.cpp

```diff
@@ -803,150 +803,179 @@ void CMobEntity::DropItems()
 
-            DropList_t* DropList = itemutils::GetDropList(m_DropID);
-            //ShowDebug(CL_CYAN"DropID: %u dropping with TH Level: %u\n" CL_RESET, PMob->m_DropID, PMob->m_THLvl);
-
-            if (DropList != nullptr && !getMobMod(MOBMOD_NO_DROPS) && DropList->Items.size() || DropList->Groups.size())
+            // check for gil (beastmen drop gil, some NMs drop gil)
+            if (CanDropGil() || (map_config.all_mobs_gil_bonus > 0 && getMobMod(MOBMOD_GIL_MAX) >= 0)) // Negative value of MOBMOD_GIL_MAX is used to prevent gil drops in Dynamis/Limbus.
             {
-                //THLvl is the number of 'extra chances' at an item. If the item is obtained, then break out.
-                uint8 maxRolls = 1 + (m_THLvl > 2 ? 2 : m_THLvl);
-                uint8 bonus = (m_THLvl > 2 ? (m_THLvl - 2) * 10 : 0);
+                charutils::DistributeGil(PChar, this); // TODO: REALISATION MUST BE IN TREASUREPOOL
+            }
 
-                for (const DropGroup_t&  group : DropList->Groups)
-                {
-                    for (uint8 roll = 0; roll < maxRolls; ++roll)
-                    {
-                        //Determine if this group should drop an item
-                        if (group.GroupRate > 0 && dsprand::GetRandomNumber(1000) < group.GroupRate * map_config.drop_rate_multiplier + bonus)
-                        {
-                            //Each item in the group is given its own weight range which is the previous value to the previous value + item.DropRate
-                            //Such as 2 items with drop rates of 200 and 800 would be 0-199 and 200-999 respectively
-                            uint16 previousRateValue = 0;
-                            uint16 itemRoll = dsprand::GetRandomNumber(1000);
-                            for (const DropItem_t& item : group.Items)
-                            {
-                                if (previousRateValue + item.DropRate > itemRoll)
-                                {
-                                    PChar->PTreasurePool->AddItem(item.ItemID, this);
-                                    break;
-                                }
-                                previousRateValue += item.DropRate;
-                            }
-                            break;
-                        }
-                    }
-                }
+            DropItems(PChar);
+        }
+
+        PChar->setWeaponSkillKill(false);
+        StatusEffectContainer->KillAllStatusEffect();
 
-                for (const DropItem_t& item : DropList->Items)
+        // NOTE: this is called for all alliance / party members!
+        luautils::OnMobDeath(this, PChar);
+
+    }
+    else
+    {
+        luautils::OnMobDeath(this, nullptr);
+    }
+}
+
+void CMobEntity::DropItems(CCharEntity* PChar) {
+    //Adds an item to the treasure pool and returns true if the pool has been filled
+    auto AddItemToPool = [this, PChar](uint16 ItemID, uint8 dropCount) -> bool {
```

Comment:
you don't need the `-> bool` part

Reply:
Ah thanks, this is my first foray into c++ lambdas - I've been stuck using pre-c++11 code. I didn't know the return value would be inferred.